Kernel: Eternal heap improvements

supercomputer7 opened the pull request:
cc @BenWiederhake

BenWiederhake commented on Kernel/Memory/MemoryManager.cpp:
```diff
@@ -49,6 +49,8 @@ namespace Kernel::Memory {
 // run. If we do, then Singleton would get re-initialized, causing
 // the memory manager to be initialized twice!
 static MemoryManager* s_the;
+
+//static bool s_initialized = false;
```
I think this should be deleted?

BenWiederhake commented on Kernel/Random.h:
```diff
@@ -121,10 +121,11 @@ class FortunaPRNG {
 
 public:
     KernelRng();
+    ~KernelRng() = delete;
```
If we give `AK_INDESTRUCTIBLE` the classname as parameter, we could make `~ClassName() = delete;` part of its definition. This would break PTYMultiplexer, but I think all the other users of this macro would benefit from it. (By which I mean: Less repetition, fewer opportunities to screw it up.)

supercomputer7 replied:
It would also break classes like `MemoryDevice` because you can't declare the destructor as deleted if it's a virtual destructor as far as I tried to do that yesterday.

BenWiederhake commented on Kernel/Heap/kmalloc.cpp:
```diff
@@ -30,7 +30,7 @@ static constexpr size_t CHUNK_SIZE = 64;
 
 #define POOL_SIZE (2 * MiB)
-#define ETERNAL_RANGE_SIZE (4 * MiB)
+#define ETERNAL_RANGE_SIZE (256 * KiB)
```
Mention in the commit description what you observed (I think you said something about how we currently only need 31 KiB?)
That makes it obvious why you think that 256 KiB is a good idea.

BenWiederhake reviewed (commented):
Looks good so far, with some suggestions for improvement.

Things I'd like someone else to think about:
- Why is PTYMultiplexer eternal/indestructible in the first place? Can't the user just create and destroy pseudo-terminals at will? As such, shouldn't the Kernel be able to reclaim that memory?
- Is the initialization re-ordering really *really* okay?
- Is the symbol-resolvability trade-off okay? (symbols become available only slightly later vs significantly more eternal heap necessary)

supercomputer7 commented:
Rebased & pushed after applying @BenWiederhake suggestions.

cc @awesomekling

BenWiederhake commented on Kernel/init.cpp:
```diff
@@ -180,11 +180,11 @@ extern "C" [[noreturn]] UNMAP_AFTER_INIT void init(BootInfo const& boot_info)
 
     load_kernel_symbol_table();
 
+    CommandLine::initialize();
```
Maybe squash this into the "Don't use eternal heap for kernel symbol data" commit.

supercomputer7 replied:
Done.

supercomputer7 commented:
I applied the last suggestion of @BenWiederhake 
This is waiting for @linusg to test this on his bare metal hardware setup and then we can continue the process of reviewing and merging :)

supercomputer7 commented:
I rebased this pull request once again. I also changed the first commit message to fix a grammar issue.

> Things I'd like someone else to think about:
> 
>     * Why is PTYMultiplexer eternal/indestructible in the first place? Can't the user just create and destroy pseudo-terminals at will? As such, shouldn't the Kernel be able to reclaim that memory?
> 
>     * Is the initialization re-ordering really _really_ okay?
> 
>     * Is the symbol-resolvability trade-off okay? (symbols become available only slightly later vs significantly more eternal heap necessary)

I know you want someone else to answer this, but let me try to do that as well:
1. the `PTYMultiplexer` device is not a device you want to de-allocate at runtime under normal circumstances, therefore it's OK to mark it as eternal. How else a user could create a pseudo terminal without it?
2. Yes. The kernel should start by initializing memory structures to ensure we can expand further before doing anything else. It also makes perfect sense to initialize the internal slab allocators, then the eternal heap (for kmalloc_eternal) and then the full-blown MemoryManager. Then the kernel is ready to actually boot the entire OS.
3. I think so. Otherwise we are doomed to calculate the kernel symbol data from time to time and expand the eternal heap - this is gross and I expect that we won't see a problem in that aspect if we apply this patch. Think about it this way - before calling the kernel symbol data initialization, we don't have kernel symbols, but the stack trace is still displayed if something wrong happened. Since we don't do a lot anyway in the init call of MemoryManager, I don't see that as a problem, and really, it's not like it moves half across the file - we move just a call or two after but then we ensure we don't have any problems with expanding the kernel in the future. It's really about solving a problem (dynamic kernel symbols and expanding of them because we add more code in the Kernel) with the right tools (use an expandable heap).
However, if you (or someone else) comes up with a decent way to calculate the amount of storage needed for kernel symbols in compile time then we can talk about it. For now that's not the case, so let's not deal with that.

BenWiederhake commented:
> > Things I'd like someone else to think about:
> 
> I know you want someone else to answer this, but let me try to do that as well:

Excellent! And sorry for my bad phrasing, I didn't intend to say "someone other than you", I actually meant "someone other than *me*, cause I don't feel confident enough on these aspects". I probably should have written something like "I can't verify these things:" or something.

> 1. the `PTYMultiplexer` device is not a device you want to de-allocate at runtime under normal circumstances, therefore it's OK to mark it as eternal. **How else a user could create a pseudo terminal without it?**

Classic example of me being unknowledgeable on that topic, I thought that somehow we need one multiplexer per PTY, and totally failed to see the line `static PTYMultiplexer& the();`.
Thanks for the explanation :)

> 2. Yes. **The kernel should start by initializing memory structures to ensure we can expand further** before doing anything else. It also makes perfect sense to initialize the internal slab allocators, then the eternal heap (for kmalloc_eternal) and then the full-blown MemoryManager. Then the kernel is ready to actually boot the entire OS.

Sounds indeed reasonable!

(Wouldn't it be nice if the language could check that certain functions *can't* be called from specific functions? That would verify, at compile-time, that the MemoryManager will *never* end up accessing the command-line or debug-symbols. Obviously, this is far, *far* out of scope, but maybe someone else reads this and feels inspired to do more static analysis.)

> 3. I think so. Otherwise we are doomed to calculate the kernel symbol data from time to time and expand the eternal heap - this is gross and I expect that we won't see a problem in that aspect if we apply this patch. Think about it this way - before calling the kernel symbol data initialization, we don't have kernel symbols, but the stack trace is still displayed if something wrong happened. Since we don't do a lot anyway in the init call of MemoryManager, I don't see that as a problem, and really, it's not like it moves half across the file - we move just a call or two after but then we ensure we don't have any problems with expanding the kernel in the future. **It's really about solving a problem (dynamic kernel symbols and expanding of them because we add more code in the Kernel) with the right tools (use an expandable heap).**
>        However, if you (or someone else) comes up with a decent way to calculate the amount of storage needed for kernel symbols in compile time then we can talk about it. In such case, you still don't allocate on the eternal heap, but you use an eternal (and fixed-size) storage for these kernel symbols to be in (like what I did with the slab allocators storage).
>        For now that's not the case, so let's not deal with that.

:+1: Makes sense to me.

> I know you ask the right questions here, but I don't think this patch needs to be overly examined... it's not like I made huge changes to how things work so I suggest to stick to the code and not to over-worry about hypothetical scenarios :)

Absolutely correct! Again, what I originally wanted to write was more along the lines of "LGTM, here are the things that I can't verify on my own, cause I'm stupid".

linusg reviewed (commented):
Tested on my usual hardware, sadly doesn't appear to be working - full boot log:

```
0.000 [Kernel]: MM: Multiboot mmap: address=0x00000000, length=644096, type=1
0.000 [Kernel]: MM: Got an unaligned physical_region from the bootloader; correcting length 644096 by 1024 bytes
0.000 [Kernel]: MM: Multiboot mmap: address=0x0009d400, length=11264, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0x000e0000, length=131072, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0x00100000, length=164102144, type=1
0.000 [Kernel]: MM: Multiboot mmap: address=0x09d80000, length=2621440, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0x0a000000, length=3542827008, type=1
0.000 [Kernel]: MM: Multiboot mmap: address=0xdd2b4000, length=1150976, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xdd3cd000, length=995328, type=1
0.000 [Kernel]: MM: Multiboot mmap: address=0xdd4c0000, length=4005888, type=4
0.000 [Kernel]: MM: Multiboot mmap: address=0xdd892000, length=11890688, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xde3e9000, length=12677120, type=1
0.000 [Kernel]: MM: Multiboot mmap: address=0xdf000000, length=16777216, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xf8000000, length=67108864, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfdf00000, length=1048576, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfea00000, length=65536, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfeb80000, length=532480, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfec10000, length=4096, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfec30000, length=4096, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfed00000, length=4096, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfed40000, length=20480, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfed80000, length=65536, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfedc2000, length=57344, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfedd4000, length=8192, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xfee00000, length=1048576, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0xff000000, length=16777216, type=2
0.000 [Kernel]: MM: Multiboot mmap: address=0x100000000, length=30588534784, type=1
0.000 [Kernel]: MM: Contiguous reserved range from P000000000009d400, length is 404480
0.000 [Kernel]: MM: Contiguous reserved range from P0000000009d80000, length is 2621440
0.000 [Kernel]: MM: Contiguous reserved range from P00000000dd2b4000, length is 1150976
0.000 [Kernel]: MM: Contiguous reserved range from P00000000dd892000, length is 11890688
0.000 [Kernel]: MM: Contiguous reserved range from P00000000df000000, length is 553648128
0.000 [Kernel]: MM: boot_pdpt @ P000000000010f000
0.000 [Kernel]: MM: boot_pd0 @ P0000000000110000
0.000 [Kernel]: MM: boot_pd_kernel @ P0000000000131000
0.000 [Kernel]: MM: Physical page entries: 0xc1e01000 - 0xc5efafff (size 0x040fa000)
0.000 [Kernel]: MM: Low memory range @ P0000000000000000 - P00000000000fffff (size 0x100000)
0.000 [Kernel]: MM: Prekernel range @ P0000000000100000 - P0000000000153fff (size 0x54000)
0.000 [Kernel]: MM: Kernel range @ P0000000000200000 - P0000000001d3dfff (size 0x1b3e000)
0.000 [Kernel]: MM: Physical Pages range @ P0000000001d3f000 - P0000000005e59fff (size 0x411b000)
0.000 [Kernel]: MM: Super physical region: P0000000000201000 - P0000000000600fff (size 0x400000)
0.000 [Kernel]:  * 4x PhysicalZone (1 MiB) @ 0000000000201000-0000000000500fff
0.000 [Kernel]: MM: User physical region: P0000000000155000 - P00000000001fefff (size 0xaa000)
0.000 [Kernel]: MM: User physical region: P0000000005e5a000 - P0000000009d7ffff (size 0x3f26000)
0.000 [Kernel]:  * 3x PhysicalZone (16 MiB) @ 0000000005e5a000-0000000007e59fff
0.000 [Kernel]:  * 15x PhysicalZone (1 MiB) @ 0000000008e5a000-0000000009c59fff
0.000 [Kernel]: MM: User physical region: P000000000a000000 - P00000000dd2b3fff (size 0xd32b4000)
0.000 [Kernel]:  * 211x PhysicalZone (16 MiB) @ 000000000a000000-00000000dbffffff
0.000 [Kernel]:  * 2x PhysicalZone (1 MiB) @ 00000000dd000000-00000000dd0fffff
0.000 [Kernel]: MM: User physical region: P00000000dd3cd000 - P00000000dd4bffff (size 0xf3000)
0.000 [Kernel]: MM: User physical region: P00000000de3e9000 - P00000000deffffff (size 0xc17000)
0.000 [Kernel]:  * 12x PhysicalZone (1 MiB) @ 00000000de3e9000-00000000deee8fff
0.000 [Kernel]: MM: User physical region: P0000000100000000 - P000000081f37ffff (size 0x1f380000)
[Kernel]: ASSERTION FAILED: s_next_eternal_ptr < s_end_of_eternal_range
[Kernel]: ../../Kernel/Heap/kmalloc.cpp:248 in void* kmalloc_eternal(size_t)
[Kernel]: KERNEL PANIC! :^(
[Kernel]: Aborted
[Kernel]: at ../../Kernel/Arch/x86/common/CPU.cpp:35 in void abort()
[Kernel]: 0xc09c9df2 (next: 0xc010c7c8)
[Kernel]: 0xc0b912e5 (next: 0xc010c808)
[Kernel]: 0xc0b910a1 (next: 0xc010c898)
[Kernel]: 0xc0c29224 (next: 0xc010c8e8)
[Kernel]: 0xc0903911 (next: 0xc010c9c8)
[Kernel]: 0xc09046ff (next: 0xc010ca18)
[Kernel]: 0xc0caa136 (next: 0xc010cbc8)
[Kernel]: 0xc0cab3c0 (next: 0xc010cc58)
[Kernel]: 0xc0cabef2 (next: 0xc010cc98)
[Kernel]: 0xc0ccf295 (next: 0xc010cd18)
[Kernel]: 0x001029b7 (next: 0x0010cff8)
[Kernel]: 0x001019fa (next: 0x00000000)
[Kernel]: ASSERTION FAILED: s_the
[Kernel]: ../../Kernel/CommandLine.cpp:32 in const Kernel::CommandLine& Kernel::kernel_command_line()
[Kernel]: KERNEL PANIC! :^(
[Kernel]: Aborted
[Kernel]: at ../../Kernel/Arch/x86/common/CPU.cpp:35 in void abort()
[Kernel]: 0xc09c9df2 (next: 0xc010c648)
[Kernel]: 0xc0b912e5 (next: 0xc010c688)
[Kernel]: 0xc0b910a1 (next: 0xc010c718)
[Kernel]: 0xc0661395 (next: 0xc010c748)
[Kernel]: 0xc09c9df7 (next: 0xc010c7c8)
[Kernel]: 0xc0b912e5 (next: 0xc010c808)
[Kernel]: 0xc0b910a1 (next: 0xc010c898)
[Kernel]: 0xc0c29224 (next: 0xc010c8e8)
[Kernel]: 0xc0903911 (next: 0xc010c9c8)
[Kernel]: 0xc09046ff (next: 0xc010ca18)
[Kernel]: 0xc0caa136 (next: 0xc010cbc8)
[Kernel]: 0xc0cab3c0 (next: 0xc010cc58)
[Kernel]: 0xc0cabef2 (next: 0xc010cc98)
[Kernel]: 0xc0ccf295 (next: 0xc010cd18)
[Kernel]: 0x001029b7 (next: 0x0010cff8)
[Kernel]: 0x001019fa (next: 0x00000000)
[Kernel]: ASSERTION FAILED: s_the
[Kernel]: ../../Kernel/CommandLine.cpp:32 in const Kernel::CommandLine& Kernel::kernel_command_line()
[Kernel]: KERNEL PANIC! :^(
[Kernel]: Aborted
[Kernel]: at ../../Kernel/Arch/x86/common/CPU.cpp:35 in void abort()
[Kernel]: 0xc09c9df2 (next: 0xc010c4c8)
[Kernel]: 0xc0b912e5 (next: 0xc010c508)
[Kernel]: 0xc0b910a1 (next: 0xc010c598)
[Kernel]: 0xc0661395 (next: 0xc010c5c8)
[Kernel]: 0xc09c9df7 (next: 0xc010c648)
[Kernel]: 0xc0b912e5 (next: 0xc010c688)
[Kernel]: 0xc0b910a1 (next: 0xc010c718)
[Kernel]: 0xc0661395 (next: 0xc010c748)
[Kernel]: 0xc09c9df7 (next: 0xc010c7c8)
[Kernel]: 0xc0b912e5 (next: 0xc010c808)
[Kernel]: 0xc0b910a1 (next: 0xc010c898)
[Kernel]: 0xc0c29224 (next: 0xc010c8e8)
[Kernel]: 0xc0903911 (next: 0xc010c9c8)
[Kernel]: 0xc09046ff (next: 0xc010ca18)
[Kernel]: 0xc0caa136 (next: 0xc010cbc8)
[Kernel]: 0xc0cab3c0 (next: 0xc010cc58)
[Kernel]: 0xc0cabef2 (next: 0xc010cc98)
[Kernel]: 0xc0ccf295 (next: 0xc010cd18)
[Kernel]: 0x001029b7 (next: 0x0010cff8)
[Kernel]: 0x001019fa (next: 0x00000000)
... Spams this indefinitely ...
```

(that last part is still #7334)

Happy to do more testing with changed patches if this gives you an idea what might be going wrong :)

Just as a side note: I know initializing the MM before everything else now was a concious decision, but I did find this output helpful before (especially printing the command line) which is now missing due to a crash in the MM.

```
0.000 [Kernel]: Loading kernel symbol table...
0.000 [Kernel]: CPU[0]: Supported features: nx pae pge rdrand rdseed smap smep sse tsc rdtscp constant_tsc nonstop_tsc sep syscall mmx sse2 sse3 ssse3 sse4.1 sse4.2 xsave avx fxsr lm
0.000 [Kernel]: CPU[0]: Physical address bit width: 48
0.000 [Kernel]: CPU[0]: Virtual address bit width: 48
0.000 [#0 Kernel]: Initializing unhandled interrupt handlers
0.000 [Kernel]: Kernel Commandline: fbdev=off serial_debug root=/dev/hda1
```

supercomputer7 commented:
I need to look into a good strategy on how to make sure we can expand physical zone data over many heaps without making it too much complicated. Closed for now.
